File: eventmesh-server-go/pkg/connector/standalone/subscribe.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package standalone

import (
	"context"
	"time"

	"github.com/apache/incubator-eventmesh/eventmesh-server-go/log"
	"github.com/apache/incubator-eventmesh/eventmesh-server-go/pkg/connector"
	"go.uber.org/atomic"
)

// ExtensionOffset key for extention about offset in cloudevent
var ExtensionOffset = "offset"

// SubscribeProcessor procese to handle the subscribe message
type SubscribeProcessor struct {
	topicName string
	broker    *Broker
	listner   connector.EventListener
	offset    *atomic.Int64
	ctx       context.Context
	cancel    context.CancelFunc
}

// NewSubscribeProcessor create new topic subscribe processer
// and start a loop to handle the message
func NewSubscribeProcessor(ctx context.Context, topicName string, br *Broker, lis connector.EventListener) *SubscribeProcessor {
	cctx, cancel := context.WithCancel(ctx)
	s := &SubscribeProcessor{
		topicName: topicName,
		broker:    br,
		listner:   lis,
		ctx:       cctx,
		cancel:    cancel,
	}
	go s.RunLoop()
	return s
}

// RunLoop run loop to process message
func (s *SubscribeProcessor) RunLoop() {
	log.Infof("process subscribe for topic:%s", s.topicName)

	tick := time.NewTicker(time.Second)
	for {
		select {
		case <-s.ctx.Done():
			return
		case <-tick.C:
			if s.offset == nil {
				msg, err := s.broker.GetMessage(s.topicName)
				if err != nil {
					log.Warnf("get message from broker err:%v, topic:%v", err, s.topicName)
					break
				}
				if msg != nil {
					if val, ok := msg.Extensions()[ExtensionOffset]; ok {
						s.offset = atomic.NewInt64(val.(int64))
					} else {
						s.offset = atomic.NewInt64(0)
					}
				}
			}

			msg, err := s.broker.GetMessageByOffset(s.topicName, s.offset.Load())
			if err != nil {
				log.Warnf("get message by offset from broker err:%v, topic:%v", err, s.topicName)
				break
			}
			s.listner.Consume(msg, func(ac connector.EventMeshAction) {
				switch ac {
				case connector.CommitMessage:
					log.Infof("commit message:%v, topic:%s, offset:%v", msg.ID(), s.topicName, s.offset.Load())
				case connector.ManualAck:
					log.Infof("manual message:%v, topic:%s, offset:%v", msg.ID(), s.topicName, s.offset.Load())
				case connector.ReconsumeLater:
					log.Infof("ack message:%v, topic:%s, offset:%v", msg.ID(), s.topicName, s.offset.Inc())
				}
			})
		}
	}
}

// Shutdown
func (s *SubscribeProcessor) Shutdown() {
	log.Infof("shutdown subscribe processor, topic:%s", s.topicName)
	s.cancel()
}

File: eventmesh-storage-plugin/eventmesh-storage-standalone/src/main/java/org/apache/eventmesh/storage/standalone/broker/task/Subscribe.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.eventmesh.storage.standalone.broker.task;

import org.apache.eventmesh.api.EventListener;
import org.apache.eventmesh.api.EventMeshAction;
import org.apache.eventmesh.api.EventMeshAsyncConsumeContext;
import org.apache.eventmesh.storage.standalone.broker.StandaloneBroker;
import org.apache.eventmesh.storage.standalone.broker.model.MessageEntity;


import io.cloudevents.CloudEvent;

import com.lmax.disruptor.EventHandler;
import com.lmax.disruptor.WorkHandler;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class Subscribe implements WorkHandler<MessageEntity>, EventHandler<MessageEntity> {

    @Getter
    private final String topicName;
    private final StandaloneBroker standaloneBroker;
    private final EventListener listener;
    @Getter
    private volatile boolean isRunning;

    public Subscribe(String topicName,
        StandaloneBroker standaloneBroker,
        EventListener listener) {
        this.topicName = topicName;
        this.standaloneBroker = standaloneBroker;
        this.listener = listener;
        this.isRunning = true;
    }

    public void subscribe() {
        standaloneBroker.subscribed(topicName, this);
    }

    public void shutdown() {
        isRunning = false;
        standaloneBroker.deleteTopicIfExist(topicName);
    }

    @Override
    public void onEvent(MessageEntity event, long sequence, boolean endOfBatch) {
        onEvent(event);
    }

    @Override
    public void onEvent(MessageEntity event) {
        try {
            if (!isRunning) {
                return;
            }
            CloudEvent message = event.getMessage();
            if (message != null) {
                EventMeshAsyncConsumeContext consumeContext = new EventMeshAsyncConsumeContext() {

                    @Override
                    public void commit(EventMeshAction action) {
                        switch (action) {
                            case CommitMessage:
                                // update offset
                                log.info("message commit, topic: {}, current offset:{}", topicName, event.getOffset());
                                break;
                            case ManualAck:
                                // update offset
                                log.info("message ack, topic: {}, current offset:{}", topicName, event.getOffset());
                                break;
                            case ReconsumeLater:
                            default:
                        }
                    }
                };
                listener.consume(message, consumeContext);
            }
        } catch (Exception ex) {
            log.error("consumer error, topic: {}, offset: {}", topicName, event.getOffset(), ex);
        }
    }

}